### PR TITLE
prov/mrail: Add raw memory region capability.

### PR DESCRIPTION
**Pull request by yburette**

### CODE REVIEW

File: prov/mrail/src/mrail_domain.c [hunk 1]
```diff
@@ -52,27 +52,167 @@ static int mrail_domain_close(fid_t fid)
...
+                                struct fi_mr_map_raw *map)
+{
+	struct mrail_mr_map_raw *mr_map;
+
+	mr_map = calloc(1, sizeof(*mr_map));
+	if (!mr_map) {
+		return -FI_ENOMEM;
+	}
+
+	mr_map->rkeys = calloc(1, map->key_size);
```
Comment (shefty) [inline prov/mrail/src/mrail_domain.c]: You can merge this into a single allocation.

File: prov/mrail/src/mrail.h [hunk 2]
```diff
@@ -106,6 +106,18 @@ struct mrail_ep {
...
+struct mrail_mr {
+	struct fid_mr mr_fid;
+	struct fid_mr **mrs;
+	size_t num_mrs;
+	uint64_t base_addr;
+};
+
+struct mrail_mr_map_raw {
+	uint64_t base_addr;
+	uint64_t *rkeys;
```
Comment (shefty) [inline prov/mrail/src/mrail.h]: Do you need the number of keys here?

File: prov/mrail/src/mrail_ep.c [hunk 3]
```diff
@@ -72,6 +78,23 @@ static ssize_t mrail_ep_sendmsg(struct fid_ep *ep_fid, const struct fi_msg *msg,
 	rail_msg.addr = *(fi_addr_t *)ofi_av_get_addr(mrail_ep->util_ep.av,
 						      (int)msg->addr);
 
+	if (rail_msg.desc) {
+		int i;
+		void **descs = calloc(rail_msg.iov_count, sizeof(void *));
```
Comment (shefty) [inline prov/mrail/src/mrail_ep.c]: This is a memory leak, plus we really don't want memory allocations in the middle if sendmsg.

File: prov/mrail/src/mrail_ep.c [hunk 4]
```diff
@@ -47,11 +47,17 @@ static ssize_t mrail_recv(struct fid_ep *ep_fid, void *buf, size_t len,
 	struct mrail_ep *mrail_ep = container_of(ep_fid, struct mrail_ep,
 					     util_ep.ep_fid.fid);
 	uint32_t rail = mrail_get_rx_rail(mrail_ep);
+	void *rail_desc = NULL;
```
Comment (shefty) [inline prov/mrail/src/mrail_ep.c]: drop the force initialization to null

File: prov/mrail/src/mrail_ep.c [hunk 5]
```diff
@@ -47,11 +47,17 @@ static ssize_t mrail_recv(struct fid_ep *ep_fid, void *buf, size_t len,
...
 	ssize_t ret;
 
 	assert(!src_addr);
 
-	ret = fi_recv(mrail_ep->eps[rail], buf, len, desc, 0, context);
+	if (desc) {
+		struct mrail_mr *mrail_mr = desc;
+		rail_desc = fi_mr_desc(mrail_mr->mrs[rail]);
+	}
```
Comment (shefty) [inline prov/mrail/src/mrail_ep.c]: Move this into a static inline function or macro that either returns the underlying descriptor or null if none is needed.

File: prov/mrail/src/mrail_ep.c [hunk 6]
```diff
@@ -89,12 +112,18 @@ static ssize_t mrail_send(struct fid_ep *ep_fid, const void *buf, size_t len,
...
 	ssize_t ret;
 
 	assert(rail_fi_addr);
 
-	ret = fi_send(mrail_ep->eps[rail], buf, len, desc, rail_fi_addr[rail],
-		      context);
+	if (desc) {
+		struct mrail_mr *mrail_mr = desc;
+		rail_desc = fi_mr_desc(mrail_mr->mrs[rail]);
+	}
```
Comment (shefty) [inline prov/mrail/src/mrail_ep.c]: see comments above for recv call

File: prov/mrail/src/mrail_domain.c [hunk 7]
```diff
@@ -52,27 +52,167 @@ static int mrail_domain_close(fid_t fid)
...
+	mrail_mr->mr_fid.fid.context = context;
+	mrail_mr->mr_fid.fid.ops = &mrail_mr_ops;
+	mrail_mr->mr_fid.mem_desc = mrail_mr;
+	mrail_mr->mr_fid.key = fi_mr_key(mrail_mr->mrs[0]);
+	mrail_mr->num_mrs = mrail_domain->num_domains;
+	*mr = &mrail_mr->mr_fid;
+
+	return 0;
+err2:
+	for (; rail!=0; --rail) {
```
Comment (shefty) [inline prov/mrail/src/mrail_domain.c]: nit: here and elsewhere, add white space before/after "=", "!=", "<", etc.

**Comment by a-ilango** [comment]
We would need to set `mr_key_size` in domain_attr in returned fi_info based on # of rails.

**Comment by yburette** [comment]
Things have changed a bit in master since my last commits. Support for FI_MR_LOCAL will come in a separate PR.

**Pull request merged**